symsrv-fetch.py

```diff
@@ -24,345 +24,468 @@
+    proc = await asyncio.create_subprocess_shell(
+        cmd, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE
+    )
+    _, err = await proc.communicate()
+    err = err.decode().strip()
+
+    return err
+
+
+async def dump_module(
+    output, symcache, filename, debug_id, code_file, code_id, has_code, dump_syms
+):
+    sym_path = os.path.join(filename, debug_id, filename.replace(".pdb", "") + ".sym")
+    output_path = os.path.join(output, sym_path)
+    sym_srv = SYM_SRV.format(symcache)
+
+    if has_code:
+        cmd = f"{dump_syms} {code_file} --code-id {code_id} --store {output} --symbol-server '{sym_srv}' --verbose error"
+    else:
+        cmd = f"{dump_syms} {filename} --debug-id {debug_id} --store {output} --symbol-server '{sym_srv}' --verbose error"
+
+    err = await run_command(cmd)
+
+    if err:
+        log.error(f"Error with {cmd}")
+        log.error(err)
+        return 1
+
+    if not has_code and not await check_x86_file(output_path):
+        log.debug(f"x86_64 binary {code_file}/{code_id} required")
+        return 2
+
+    log.info(f"Successfully dumped: {filename}/{debug_id}")
+    return sym_path
+
+
+async def dump_helper(output, symcache, modules, dump_syms):
+    tasks = []
+    for filename, debug_id, code_file, code_id, has_code in modules:
+        tasks.append(
+            dump_module(
+                output,
+                symcache,
+                filename,
+                debug_id,
+                code_file,
+                code_id,
+                has_code,
+                dump_syms,
+            )
+        )
+
+    return await asyncio.gather(*tasks)
+
+
+def dump(output, symcache, modules, dump_syms):
+    res = asyncio.run(dump_helper(output, symcache, modules, dump_syms))
+    file_index = {x for x in res if isinstance(x, str)}
+    stats = {
+        "dump_error": sum(1 for x in res if x == 1),
+        "no_bin": sum(1 for x in res if x == 2),
+    }
+
+    return file_index, stats
+
+
+async def collect_helper(modules):
+    loop = asyncio.get_event_loop()
+    tasks = []
+    connector = TCPConnector(limit=100, limit_per_host=20)
+    async with ClientSession(
+        loop=loop, timeout=ClientTimeout(total=TIMEOUT), connector=connector
+    ) as client:
+        for filename, ids in modules.items():
+            for debug_id, code_file, code_id in ids:
+                tasks.append(
+                    collect_info(client, filename, debug_id, code_file, code_id)
+                )
+
+        return await asyncio.gather(*tasks)
+
+
+def collect(modules):
+    res = asyncio.run(collect_helper(modules))
+    to_dump = []
+    stats = {"no_pdb": 0, "is_there": 0}
+    for filename, debug_id, code_file, code_id, has_pdb, has_code, is_there in res:
+        if not has_pdb:
+            if is_there:
+                stats["is_there"] += 1
+            else:
+                stats["no_pdb"] += 1
+                log.info(f"No pdb for {filename}/{debug_id}")
```

Comment:
The stats and this message would be wrong as actually `has_pdb` can be False even though the PDB is there if the file is on Mozilla's symbol server. Do we care?